File: backend/app/engines/payment_engine.py
"""
payment_engine.py — VapeERP 결제/적립 엔진 v4 (Python)
paymentEngine.js v4와 동일한 로직.

채널별 확정 정책:
  매장:       현금·이체·카드·마일리지 모두 가능 / 적립 가능
  배달·택배:  현금·이체만 / 카드·마일리지 불가 / 적립 없음
"""

from dataclasses import dataclass
from typing import Optional
import math


CHANNEL_POLICY = {
    '매장': {'card_ok': True,  'miles_ok': True,  'earn_ok': True},
    '배달': {'card_ok': False, 'miles_ok': False, 'earn_ok': False},
    '택배': {'card_ok': False, 'miles_ok': False, 'earn_ok': False},
}


@dataclass
class PaymentResult:
    nature:          str  = ''
    card_pct:        int  = 0
    svc_eligible:    bool = False
    earn_eligible:   bool = False
    earn_reason:     str  = ''
    earn_amt:        int  = 0
    total_to_pay:    int  = 0
    miles_final:     int  = 0
    is_balance_ok:   bool = False
    card_blocked:    bool = False
    miles_blocked:   bool = False


def calc_payment(
    cash: int, card: int, miles_used: int, miles_balance: int,
    subtotal: int, liquid_price: int, device_price: int,
    services: list, channel: str,
) -> PaymentResult:

    res = PaymentResult()
    policy = CHANNEL_POLICY.get(channel, CHANNEL_POLICY['매장'])

    # 채널별 차단
    res.card_blocked  = not policy['card_ok']  and card > 0
    res.miles_blocked = not policy['miles_ok'] and miles_used > 0

    card_final  = card  if policy['card_ok']  else 0
    max_miles   = math.floor(min(miles_balance, subtotal) / 10) * 10 if policy['miles_ok'] else 0
    miles_final = math.floor(min(miles_used, max_miles) / 10) * 10  if policy['miles_ok'] else 0
    res.miles_final = miles_final

    total_to_pay = subtotal - miles_final
    res.total_to_pay = total_to_pay

    # 카드 비중
    total_input = cash + card_final + miles_final
    card_pct    = round(card_final / total_input * 100) if total_input > 0 else 0
    res.card_pct = card_pct

    # 결제 성격
    if card_final == 0 and miles_final == 0:
        nature = '현금/이체'
    elif card_final == 0:
        nature = '마일리지 전액' if miles_final >= subtotal else '현금/이체+마일리지'
    elif card_pct <= 20:
        nature = '현금/이체 성격 (카드 20% 이하)'
    else:
        nature = '카드'
    res.nature = nature

    # 서비스 자격 (배달·택배는 카드 없으므로 항상 가능)
    res.svc_eligible = not nature.startswith('카드')

    # 결제 합산 검증
    res.is_balance_ok = abs((cash + card_final + miles_final) - subtotal) <= 10

    # 적립
    has_earn_block = any(
        s.get('earnBlock', False) if isinstance(s, dict) else getattr(s, 'earn_block', False)
        for s in services
    )
    earn_base = liquid_price + device_price

    if not policy['earn_ok']:
        res.earn_reason = f'{channel} 채널 → 적립 없음'
    elif miles_final > 0:
        res.earn_reason = '마일리지 사용 → 적립 없음'
    elif has_earn_block:
        res.earn_reason = '고가팟 특수 서비스 거래 → 적립 없음'
    elif earn_base <= 0:
        res.earn_reason = '적립 대상 품목 없음 (코일·악세사리만)'
    else:
        res.earn_eligible = True
        if nature == '카드':
            res.earn_amt    = math.floor(earn_base * 0.05 / 10) * 10
            res.earn_reason = f'카드 결제 · 액상+기기 기준 5% = {res.earn_amt:,}원'
        else:
            res.earn_amt    = math.floor(total_to_pay * 0.05 / 10) * 10
            res.earn_reason = f'현금/이체 기준 5% = {res.earn_amt:,}원'

    return res

File: frontend/src/engines/paymentEngine.js
/**
 * paymentEngine.js — VapeERP 결제 성격 판정 & 적립 엔진 v4
 *
 * ── 채널별 확정 정책 ────────────────────────────────────
 * 매장:       현금·이체·카드·마일리지 모두 가능
 * 배달·택배:  현금·이체만 가능. 카드·마일리지 불가
 *
 * ── 결제 성격 판정 ──────────────────────────────────────
 * 카드 비중 = 카드금액 ÷ (현금+이체+카드+마일리지) × 100
 *   0%     → 현금/이체
 *   1~20%  → 현금/이체 성격 (서비스 가능)
 *   >20%   → 카드 (서비스 불가)
 *   전액마일리지 → 현금/이체 성격 (서비스 가능, 적립 없음, 매장만)
 *
 * ── 적립 조건 (AND) ─────────────────────────────────────
 * ① 채널 = 매장
 * ② 마일리지 사용 = 0
 * ③ 고가팟 특수 서비스 거래 아님 (earnBlock)
 * ④ 적립 대상 품목(액상·기기) 포함
 *
 * ── 적립 계산 (10원 단위 버림) ──────────────────────────
 * 현금/이체: floor(최종결제금액 × 0.05 / 10) × 10
 * 카드:      floor((액상금액+기기금액) × 0.05 / 10) × 10
 */

const CHANNEL_POLICY = {
  '매장': { cardOk: true,  milesOk: true,  earnOk: true  },
  '배달': { cardOk: false, milesOk: false, earnOk: false },
  '택배': { cardOk: false, milesOk: false, earnOk: false },
};

/**
 * calcPayment({ cash, card, milesUsed, milesBalance,
 *               priceResult, services, channel })
 */
export function calcPayment({ cash = 0, card = 0, milesUsed = 0, milesBalance = 0,
                               priceResult, services, channel }) {
  const { subtotal, liquidPrice, devicePrice } = priceResult;
  const policy = CHANNEL_POLICY[channel] ?? CHANNEL_POLICY['매장'];

  // ── 채널별 차단 ─────────────────────────────────────
  const cardBlocked  = !policy.cardOk  && card > 0;
  const milesBlocked = !policy.milesOk && milesUsed > 0;

  // 실제 적용 금액 (차단된 수단은 0 처리)
  const cardFinal  = policy.cardOk  ? (card  ?? 0) : 0;
  const maxMiles   = policy.milesOk ? Math.min(milesBalance, subtotal) : 0;
  const milesFloor = Math.floor(maxMiles / 10) * 10;
  const milesFinal = Math.min(Math.floor((milesUsed ?? 0) / 10) * 10, milesFloor);
  const cashFinal  = cash ?? 0;

  const totalToPay = subtotal - milesFinal;

  // ── 카드 비중 ────────────────────────────────────────
  const totalInput = cashFinal + cardFinal + milesFinal;
  const cardPct    = totalInput > 0 ? cardFinal / totalInput * 100 : 0;

  // ── 결제 성격 판정 ───────────────────────────────────
  let nature;
  if (cardFinal === 0 && milesFinal === 0) {
    nature = '현금/이체';
  } else if (cardFinal === 0) {
    nature = milesFinal >= subtotal ? '마일리지 전액' : '현금/이체+마일리지';
  } else if (cardPct <= 20) {
    nature = '현금/이체 성격 (카드 20% 이하)';
  } else {
    nature = '카드';
  }

  // ── 서비스 자격 ─────────────────────────────────────
  // 배달·택배는 카드 없으므로 항상 현금/이체 성격 → 서비스 가능
  const svcEligible = !nature.startsWith('카드');

  // ── 결제 합산 검증 ───────────────────────────────────
  const isBalanceOk = Math.abs((cashFinal + cardFinal + milesFinal) - subtotal) <= 10;

  // ── 적립 판정 ────────────────────────────────────────
  let earnEligible = false, earnReason = '', earnAmt = 0;
  const hasEarnBlock = (services ?? []).some(s => s.earnBlock);
  const earnBase     = liquidPrice + devicePrice;

  if (!policy.earnOk) {
    earnReason = `${channel} 채널 → 적립 없음`;
  } else if (milesFinal > 0) {
    earnReason = '마일리지 사용 → 적립 없음';
  } else if (hasEarnBlock) {
    earnReason = '고가팟 특수 서비스 거래 → 적립 없음';
  } else if (earnBase <= 0) {
    earnReason = '적립 대상 품목 없음 (코일·악세사리만)';
  } else {
    earnEligible = true;
    if (nature === '카드') {
      earnAmt    = Math.floor(earnBase * 0.05 / 10) * 10;
      earnReason = `카드 결제 · 액상+기기 기준 5% = ${earnAmt.toLocaleString()}원`;
    } else {
      earnAmt    = Math.floor(totalToPay * 0.05 / 10) * 10;
      earnReason = `현금/이체 기준 5% = ${earnAmt.toLocaleString()}원`;
    }
  }

  return {
    nature,
    cardPct:        Math.round(cardPct),
    svcEligible,
    earnEligible,
    earnReason,
    earnAmt,
    totalToPay,
    milesFinal,
    isBalanceOk,
    cardBlocked,    // UI에서 카드 입력란 비활성화 신호
    milesBlocked,   // UI에서 마일리지 입력란 비활성화 신호
  };
}
